foolbox/attacks/spatial.py

```diff
@@ -0,0 +1,130 @@
+# -*- coding: utf-8 -*-
+from __future__ import division
+
+import numpy as np
+from itertools import product
+from scipy.ndimage import rotate, shift
+import operator
+
+from .base import Attack
+from .base import call_decorator
+
+
+class SpatialAttack(Attack):
+    """Adversarially chosen rotations and translations [1]. This implementation
+    is based on the reference implementation by Madry et al.
+    https://github.com/MadryLab/adversarial_spatial
+
+    References
+    ----------
+    .. [1] Logan Engstrom*, Brandon Tran*, Dimitris Tsipras*,
+           Ludwig Schmidt, Aleksander Mądry: "A Rotation and a
+           Translation Suffice: Fooling CNNs with Simple Transformations",
+           http://arxiv.org/abs/1712.02779
+    """
+
+    @call_decorator
+    def __call__(self, input_or_adv, label=None, unpack=True,
+                 do_rotations=True, do_translations=True,
+                 x_shift_limits=(-5, 5), y_shift_limits=(-5, 5),
+                 angular_limits=(-5, 5), granularity=10,
+                 random_sampling=False, abort_early=True):
+
+        """Adversarially chosen rotations and translations.
+
+        Parameters
+        ----------
+        input_or_adv : `numpy.ndarray` or :class:`Adversarial`
+            The original, unperturbed input as a `numpy.ndarray` or
+            an :class:`Adversarial` instance.
+        label : int
+            The reference label of the original input. Must be passed
+            if `a` is a `numpy.ndarray`, must not be passed if `a` is
+            an :class:`Adversarial` instance.
+        unpack : bool
+            If true, returns the adversarial input, otherwise returns
+            the Adversarial object.
+        do_rotations : bool
+            If False no rotations will be applied to the image.
+        do_translations : bool
+            If False no translations will be applied to the image.
+        x_shift_limits : int or (int, int)
+            Limits for horizontal translations in pixels. If one integer is
+            provided the limits will be (-x_shift_limits, x_shift_limits).
+        y_shift_limits : int or (int, int)
+            Limits for vertical translations in pixels. If one integer is
+            provided the limits will be (-y_shift_limits, y_shift_limits).
+        angular_limits : int or (int, int)
+            Limits for rotations in degrees. If one integer is
+            provided the limits will be [-angular_limits, angular_limits].
+        granularity : int
+            Density of sampling within limits for each dimension.
+        random_sampling : bool
+            If True we sample translations/rotations randomly within limits,
+            otherwise we use a regular grid.
+        abort_early : bool
```

Comment:
Do we need `abort_early` (it's possible to get the same behavior by passing threshold=np.inf to `Adversarial`). On the other hand, given that it's on by default, we might still want it here. Is there actually a reason to ever disable it for this attack?